Update substring match for substitutions

faculerena opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/148061)*
<!-- homu-ignore:end -->

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

Fixed `as_substr` function to check for matching prefixes too. This fixes this issue where `td::s` is matched instead of `std::`. This same behaviour can be seen with `time` and `tokio::time` and similar cases where the error import starts with the same prefix as the suggestion.

```rust
use sync;
fn main() {}
```


```
error[E0432]: unresolved import `sync`
 --> /tmp/test_import.rs:1:5
  |
1 | use sync;
  |     ^^^^ no `sync` in the root
  |
help: consider importing this module instead
  |
1 | use std::sync;
  |      +++++
```

After the changes:

```
error[E0432]: unresolved import `sync`
 --> /tmp/test_import.rs:1:5
  |
1 | use sync;
  |     ^^^^ no `sync` in the root
  |
help: consider importing this module instead
  |
1 | use std::sync;
  |     +++++
```

Kivooeo commented:
Well, I guess we can't modify this `as_substr`, seems like even smallest change breaks everything else, 

I guess we should somehow modify this function

https://github.com/rust-lang/rust/blob/ab925646fae038b02bd462cd328ae9eef1639236/compiler/rustc_resolve/src/imports.rs#L736

This function was pointed by `-Ztrack-diagnostics`

```rust
error[E0432]: unresolved import `sync`
 --> /home/gh-Kivooeo/test_/src/main.rs:2:5
  |
2 | use sync;
  |     ^^^^ no `sync` in the root
  |
  = note: -Ztrack-diagnostics: created at compiler/rustc_resolve/src/imports.rs:768:24
help: consider importing this module instead
  |
2 | use std::sync;
  |      +++++
```

This one generates a final suggestion, but I don't really sure how this tide with `as_substr` (haven't checked stack trace yet, but I guess you did this already, otherwise I have no idea how you found this `as_substr` function)

So, what is my idea about, in the place where we generating suggestion (I guess this is function above, but I'm not 100% sure) we need to check what we got:

1. `crate::` 
2. `rate::c` or `ate::cr`

so if we got first we are fine with that, if we got second then we should somehow move `cr` to the start, so we will get `crate::`, here we have some advantage because we always know that everything after `::` was stripped from start, so it should be moved

I will be honest that I have no idea how to implement that in that function, I maybe will took another look but later, so there is just some of my thoughts on the core idea how this should work

Kivooeo commented:
```diff
diff --git a/compiler/rustc_errors/src/lib.rs b/compiler/rustc_errors/src/lib.rs
index 17cd466f96b..2ed4e281adb 100644
--- a/compiler/rustc_errors/src/lib.rs
+++ b/compiler/rustc_errors/src/lib.rs
@@ -306,6 +306,18 @@ fn replaces_meaningful_content(&self, sm: &SourceMap) -> bool {
 /// `BB` is. Return the length of the prefix, the "trimmed" suggestion, and the length
 /// of the suffix.
 fn as_substr<'a>(original: &'a str, suggestion: &'a str) -> Option<(usize, &'a str, usize)> {
+    if suggestion.contains("::")
+        && ["core", "std"].iter().any(|&prefix| suggestion.starts_with(prefix))
+        && suggestion.ends_with(original)
+        && let Some(c1) = suggestion.chars().nth(0)
+        && let Some(c2) = original.chars().nth(0)
+        && c1 == c2
+    {
+        let prefix_len = suggestion.len() - original.len();
+        let prefix = &suggestion[..prefix_len];
+        return Some((0, prefix, original.len()));
+    }
+
     let common_prefix = original
         .chars()
         .zip(suggestion.chars())
```

Btw, just decided to check will this work or not... and sadly it is, all tests passed, and it gave correct suggestion 

But... I'd consider this as a last resort if we can't come up with a better solution

faculerena commented:
Yeah, I was just trying to generalize that solution, as that does not catch third party crates. But yeah, it looks like the solution is that way.

Kivooeo commented:
I guess it's possible to remove `core` and `std` check from here, I was just lazy to rerun tests once again to test it

It's a bit late for me atm (3 am) so I would able to test it without this check later today

Or if you want to test it yourself, you can copy this into your rustc fork and run `./x test ui` you will see if there any regressions in test suite

But I'd be happier to see an implementation of my idea above, I also will try to investigate this direction more later on today

faculerena committed:
Update substringing when orig & sugg have equal prefixes

Kivooeo commented:
Have you tried to run tests locally? Because to update them you to pass `--bless` flag, which will update "expected output"

faculerena committed:
Change existing test to reflect new behaviour

The change in this test is minimal, as this is a specific case. The benefits for other cases explained in issue 148070 outweight the change.

faculerena commented:
Ok, finally I updated a test to reflect this new behaviour, as I think it's just a test quirk more than a lint change. Thanks @Kivooeo for the help too.

Kivooeo commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -245,6 +245,7 @@ pub(crate) struct SubstitutionHighlight {
 impl SubstitutionPart {
     /// Try to turn a replacement into an addition when the span that is being
     /// overwritten matches either the prefix or suffix of the replacement.
+    #[instrument(level = "debug", skip(self, sm))]
```
Why this is needed?

faculerena committed:
Remove debug `instrument` and added comment explaning the special case

Kivooeo commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -306,6 +306,19 @@ impl TrimmedSubstitutionPart {
+        if prefix.ends_with("::") && suggestion.chars().next() == original.chars().next() {
+            return Some((0, prefix, original.len()));
+        }
+    }
```
```suggestion
    if suggestion.contains("::")
        && suggestion.ends_with(original)
        && suggestion.len() > original.len()
        && let prefix = &suggestion[..suggestion.len() - original.len()]
        && prefix.ends_with("::")
        && suggestion.chars().next() == original.chars().next()
    {
        return Some((0, prefix, original.len()));
    }
```

faculerena committed:
Move `let` inside `if` condition

faculerena committed:
Added tests for correct highlighting of strings with shared prefix

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19048485643/job/54402652461) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/54402652461) [(plain)](https://github.com/rust-lang/rust/commit/fd13a61286ca95cd06739c939828efa63449e27d/checks/54402652461/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
To only update this specific test, also pass `--test-args imports/same-prefix-unresolved-import-148070.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/imports/same-prefix-unresolved-import-148070" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0432]: unresolved import `stat`
##[error]  --> /checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs:3:5
   |
LL | use stat; //~ ERROR unresolved import `stat`
   |     ^^^^ no `stat` in the root
   |
help: consider importing this struct instead
   |
LL | use std::os::linux::raw::stat; //~ ERROR unresolved import `stat`
   |     +++++++++++++++++++++

error[E0432]: unresolved import `str`
##[error]  --> /checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs:4:5
   |
LL | use str; //~ ERROR unresolved import `str`
   |     ^^^ no `str` in the root
   |
help: a similar name exists in the module
   |
LL - use str; //~ ERROR unresolved import `str`
LL + use std; //~ ERROR unresolved import `str`
   |
help: consider importing one of these items instead
   |
LL | use std::primitive::str; //~ ERROR unresolved import `str`
   |     ++++++++++++++++
LL | use std::str; //~ ERROR unresolved import `str`
   |     +++++

error[E0432]: unresolved import `sync`
##[error]  --> /checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs:5:5
   |
LL | use sync; //~ ERROR unresolved import `sync`
   |     ^^^^ no `sync` in the root
   |
help: consider importing this module instead
   |
LL | use std::sync; //~ ERROR unresolved import `sync`
   |     +++++

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0432`.
```

</details>

BoxyUwU commented:
@bors r+ rollup

certainly a special case but also does seem like a nice improvement

faculerena commented:
Don't know why it's failing there. The logs show the wrong highlighting so I guess something is out of sync?

Kivooeo commented:
As Jonathan said there is some soft conflict, so maybe something because of this, but I agree that this is weird because locally it does work correctly

@JonathanBrouwer, can we try this in other rollup?

JonathanBrouwer commented:
Lets try the `aarch64-gnu-llvm-20-1` job after https://github.com/rust-lang/rust/pull/148802 merges (so if there are any soft conflicts we'll know), and if that passes we can try again

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19245778971/job/55019387565) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/55019387565) [(plain)](https://github.com/rust-lang/rust/commit/9ffa1fb720b78b72391c3362298aef9a1b0f444c/checks/55019387565/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
To only update this specific test, also pass `--test-args imports/same-prefix-unresolved-import-148070.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/imports/same-prefix-unresolved-import-148070" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0432]: unresolved import `stat`
##[error]  --> /checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs:3:5
   |
LL | use stat; //~ ERROR unresolved import `stat`
   |     ^^^^ no `stat` in the root
   |
help: consider importing this struct instead
   |
LL | use std::os::linux::raw::stat; //~ ERROR unresolved import `stat`
   |       +++++++++++++++++++++

error[E0432]: unresolved import `str`
##[error]  --> /checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs:4:5
   |
LL | use str; //~ ERROR unresolved import `str`
   |     ^^^ no `str` in the root
   |
help: a similar name exists in the module
   |
LL - use str; //~ ERROR unresolved import `str`
LL + use std; //~ ERROR unresolved import `str`
   |
help: consider importing one of these items instead
   |
LL | use std::primitive::str; //~ ERROR unresolved import `str`
   |       ++++++++++++++++
LL | use std::str; //~ ERROR unresolved import `str`
   |       +++++

error[E0432]: unresolved import `sync`
##[error]  --> /checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs:5:5
   |
LL | use sync; //~ ERROR unresolved import `sync`
   |     ^^^^ no `sync` in the root
   |
help: consider importing this module instead
   |
LL | use std::sync; //~ ERROR unresolved import `sync`
   |      +++++

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0432`.
---
To only update this specific test, also pass `--test-args test-attrs/inaccessible-test-modules.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/test-attrs/inaccessible-test-modules.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/test-attrs/inaccessible-test-modules" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--test"
stdout: none
--- stderr -------------------------------
error[E0432]: unresolved import `main`
##[error]  --> /checkout/tests/ui/test-attrs/inaccessible-test-modules.rs:5:5
   |
LL | use main as x; //~ ERROR unresolved import `main`
   |     ^^^^^^^^^ no `main` in the root

error[E0432]: unresolved import `test`
##[error]  --> /checkout/tests/ui/test-attrs/inaccessible-test-modules.rs:6:5
   |
LL | use test as y; //~ ERROR unresolved import `test`
   |     ^^^^^^^^^ no `test` in the root
   |
help: consider importing this module instead
   |
LL | use test::test as y; //~ ERROR unresolved import `test`
   |         ++++++

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0432`.
```

</details>

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19247363386/job/55024576557) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/55024576557) [(plain)](https://github.com/rust-lang/rust/commit/a57969bcaa489c0f5864daf24649c80b9baf552e/checks/55024576557/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
To only update this specific test, also pass `--test-args imports/same-prefix-unresolved-import-148070.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/imports/same-prefix-unresolved-import-148070" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0432]: unresolved import `stat`
##[error]  --> /checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs:3:5
   |
LL | use stat; //~ ERROR unresolved import `stat`
   |     ^^^^ no `stat` in the root
   |
help: consider importing this struct instead
   |
LL | use std::os::linux::raw::stat; //~ ERROR unresolved import `stat`
   |       +++++++++++++++++++++

error[E0432]: unresolved import `str`
##[error]  --> /checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs:4:5
   |
LL | use str; //~ ERROR unresolved import `str`
   |     ^^^ no `str` in the root
   |
help: a similar name exists in the module
   |
LL - use str; //~ ERROR unresolved import `str`
LL + use std; //~ ERROR unresolved import `str`
   |
help: consider importing one of these items instead
   |
LL | use std::primitive::str; //~ ERROR unresolved import `str`
   |       ++++++++++++++++
LL | use std::str; //~ ERROR unresolved import `str`
   |       +++++

error[E0432]: unresolved import `sync`
##[error]  --> /checkout/tests/ui/imports/same-prefix-unresolved-import-148070.rs:5:5
   |
LL | use sync; //~ ERROR unresolved import `sync`
   |     ^^^^ no `sync` in the root
   |
help: consider importing this module instead
   |
LL | use std::sync; //~ ERROR unresolved import `sync`
   |      +++++

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0432`.
---
To only update this specific test, also pass `--test-args test-attrs/inaccessible-test-modules.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/test-attrs/inaccessible-test-modules.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/test-attrs/inaccessible-test-modules" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--test"
stdout: none
--- stderr -------------------------------
error[E0432]: unresolved import `main`
##[error]  --> /checkout/tests/ui/test-attrs/inaccessible-test-modules.rs:5:5
   |
LL | use main as x; //~ ERROR unresolved import `main`
   |     ^^^^^^^^^ no `main` in the root

error[E0432]: unresolved import `test`
##[error]  --> /checkout/tests/ui/test-attrs/inaccessible-test-modules.rs:6:5
   |
LL | use test as y; //~ ERROR unresolved import `test`
   |     ^^^^^^^^^ no `test` in the root
   |
help: consider importing this module instead
   |
LL | use test::test as y; //~ ERROR unresolved import `test`
   |         ++++++

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0432`.
```

</details>

faculerena commented:
Shouldn't the fix, as it is right now here, work for stable then? I don't fully understand that.

Kivooeo commented:
Not really, just because you pushing changes for main branch, and you can't change or replace things in older versions, since main branch is around 1.93 version and stable is 1.91.1, so you can't directly change it, so in this case we have to wait when stable become 1.93 and catch this changes 

(I tried to explain that, it was pretty hard for non-native speaker lol, hope it makes at least little sense)

faculerena commented:
Don't worry, I'm not either. I understood perfectly lol, thanks! 

I sent a PR to `annotate_snippets`, it's exactly the same fix as here, in the same function.

Kivooeo commented:
Given that https://github.com/rust-lang/rust/pull/148984 will fix that issue, but it does not introduce a regression test, so it would be reasonable imo for this pr add this test

SpriteOvO commented:
@faculerena Hi, ping from triage team. This PR has been inactive for a while. Based on my understanding after reading the thread, the fix has already been merged into an external crate. It would be great if you could continue this PR by adding a test for it in rustc. Thanks ^^
